src/app/clusters/scenes-server/SceneHandlerImpl.cpp

```diff
@@ -16,10 +16,156 @@
  */
 
 #include <app/clusters/scenes-server/SceneHandlerImpl.h>
+#include <app/util/endpoint-config-api.h>
 
 namespace chip {
 namespace scenes {
 
+namespace {
+
+using ConcreteAttributePath  = app::ConcreteAttributePath;
+using AttributeValuePairType = app::Clusters::ScenesManagement::Structs::AttributeValuePair::Type;
+
+/// ConvertByteArrayToUInt64
+/// @brief Helper function to convert a byte array to a uint64_t value
+/// @param EmberAfDefaultAttributeValue & defaultValue
+/// @param len Length of the byte array
+/// @return uint64_t or int64_t Value
+/// @note The attribute table supports 8 bytes only for unsigned integers, 4 bytes is the maximum for signed integers
+template <typename Type>
+Type ConvertByteArrayToType(const EmberAfDefaultAttributeValue & defaultValue, uint16_t len)
+{
+    if (len <= 2)
+    {
+        return static_cast<Type>(defaultValue.defaultValue);
+    }
+    else
+    {
+        Type sValue = 0;
+        memcpy(&sValue, defaultValue.ptrToDefaultValue, sizeof(Type));
+        return app::NumericAttributeTraits<Type>::StorageToWorking(sValue);
+    }
+}
+
+/// CapAttributeID
+/// Cap the attribute value based on the attribute's min and max if they are defined,
+/// or based on the attribute's size if they are not.
+/// @param[in] aVPair   AttributeValuePairType
+/// @param[in] metadata  EmberAfAttributeMetadata
+///
+template <typename Type>
+void CapAttributeID(AttributeValuePairType & aVPair, const EmberAfAttributeMetadata * metadata)
+{
+    // Calculate the maximum value that can be represented with the given number of bytes
+    Type maxValue = 0;
+
+    // Check if the attribute type is signed
+    if (metadata->IsSignedIntegerAttribute())
+    {
+        maxValue = static_cast<Type>((1ULL << (emberAfAttributeSize(metadata) * 8 - 1)) - 1);
+    }
+    else
+    {
+        maxValue = static_cast<Type>((1ULL << (emberAfAttributeSize(metadata) * 8)) - 1);
+    }
+
+    // Check metadata for min and max values
+    if (metadata->HasMinMax())
+    {
+        const EmberAfAttributeMinMaxValue * minMaxValue = metadata->defaultValue.ptrToMinMaxValue;
+        Type minVal = ConvertByteArrayToType<Type>(minMaxValue->minValue, emberAfAttributeSize(metadata));
+        Type maxVal = ConvertByteArrayToType<Type>(minMaxValue->maxValue, emberAfAttributeSize(metadata));
+
+        // Cap based on minimum value
+        if (minVal > static_cast<Type>(aVPair.attributeValue))
+        {
+            uint64_t sValue = 0;
+            memcpy(&sValue, &minVal, sizeof(Type));
+            aVPair.attributeValue = app::NumericAttributeTraits<uint64_t>::StorageToWorking(sValue);
+            // We assume the max is >= min therefore we can return
+            return;
+        }
+
+        // Adjust maxValue if greater than the meta data's max value
+        if (maxVal < maxValue)
+        {
+            maxValue = maxVal;
+        }
+    }
+
+    // Cap based on maximum value
+    if (metadata->IsSignedIntegerAttribute())
```

Comment:
Don't we already know this information based on Type, at compile time?  This feels pretty strange to me...

Reply:
This is because we keep the value in it's type's size inside of attributeValue,
Eg -1 for an int16_t would be stored as FFFF'0000'0000'0000 instead of FFFF'FFFF'FFFF'FFFF,
So we need to cast it as a uint64_t for the comparison only.